gpii/node_modules/eventLog/src/eventLog.js

```diff
@@ -0,0 +1,396 @@
+    }
+
+    gpii.eventLog.writeLog(level, eventObject);
+};
+
+/**
+ * Defers logging of an event until a later time (when logDeferred() is called with the same timerId, or the application
+ * exits).
+ *
+ * The purpose of this function is to prepare a duration log entry for later, but without requiring the caller to
+ * remember the details when the duration is over. This entry will be logged if GPII closes before a corresponding call
+ * of logDeferred.
+ *
+ * @param module {String} The part of GPII causing this event.
+ * @param event {String} The event name.
+ * @param data {Object} Event specific data.
+ * @param level {Object} [optional] Level of the log, see fluid.logLevelsSpec [FATAL,FAIL,WARN,IMPORTANT,INFO,TRACE].
+ * @param timerId {String} The timerId.
+ */
+gpii.eventLog.defer = function (module, event, data, level, timerId) {
+    if (!timerId) {
+        timerId = level;
+        level = null;
+    }
+    gpii.eventLog.startTimer(timerId);
+    gpii.eventLog.pendingLogs[timerId] = gpii.eventLog.createLogObject(module, event, data, level);
+};
+
+/**
+ * Logs a deferred entry, setting the duration field to the data object.
+ *
+ * @param timerId {String} The same timerId for the call to defer().
+ */
+gpii.eventLog.logDeferred = function (timerId) {
+    var eventObject = gpii.eventLog.pendingLogs[timerId];
+    if (eventObject) {
+        delete gpii.eventLog.pendingLogs[timerId];
+
+        gpii.eventLog.logDuration(eventObject, null, null, null, timerId);
+    } else {
+        fluid.log("logDeferred: Unknown timerId", timerId);
+
+    }
+};
+
+/**
+ * Logs an error.
+ *
+ * @param module {String} The part of GPII causing this error.
+ * @param errType {String} Type of error.
+ * @param err {Object} The error.
+ */
+gpii.eventLog.logError = function (module, errType, err) {
+    var data = {};
+    if (err instanceof Error) {
+        // Error doesn't serialise
+        data.error = {};
+        fluid.each(Object.getOwnPropertyNames(err), function (a) {
+            data.error[a] = err[a];
+        });
+    } else {
+        data.error = err;
+    }
+    gpii.eventLog.log(module, "Error." + errType, data, fluid.logLevel.FAIL);
+};
+
+// Log uncaught exceptions.
+fluid.onUncaughtException.addListener(function (err) {
+    gpii.eventLog.logError(null, "UncaughtException", err);
+}, "eventLog");
+
+// Log fluid.fail.
+fluid.failureEvent.addListener(function (args) {
+    var err = Array.isArray(args) ? args.join(" ") : err;
+    gpii.eventLog.logError(null, "Failure", err);
+}, "eventLog", "before:fail");
+
+/**
+ * Gets the path of the new log file for this instance of gpii.
+ *
+ * @return {String} Path of the log file.
+ */
+gpii.eventLog.getLogFilePath = function () {
+    if (!gpii.eventLog.logFilePath) {
+        var startupTime = Date.now();
+        gpii.eventLog.logFilePath = process.env.GPII_EVENT_LOG;
+
+        if (!gpii.eventLog.logFilePath) {
+            var settingsDirComponent = gpii.settingsDir();
```

Comment:
Make this a proper subcomponent, and this function into an invoker which references it. Standalone Infusion components have awkward implications, and will leak, for example